bigtable-dataflow-parent/bigtable-hbase-dataflow/src/test/java/com/google/cloud/bigtable/dataflow/CloudBigtableIOReaderTest.java

```diff
@@ -78,4 +80,53 @@ void initializeScanner() throws IOException {
 
     underTest.close();
   }
+
+  private void setRowKey(String rowKey) throws IOException {
+    ByteString rowKeyByteString = ByteString.copyFrom(Bytes.toBytes(rowKey));
+    Row row = Row.newBuilder().setKey(rowKeyByteString).build();
+    when(mockScanner.next()).thenReturn(row);
+  }
+
+  private CloudBigtableIO.Reader<Result> initializeReader(CloudBigtableScanConfiguration config) {
+    when(mockSource.getConfiguration()).thenReturn(config);
+    return new CloudBigtableIO.Reader<Result>(mockSource, CloudBigtableIO.RESULT_ADVANCER) {
+      @Override
+      void initializeScanner() throws IOException {
+        setSession(mockSession);
+        setScanner(mockScanner);
+      }
+    };
+  }
+
+  @Test
+  public void testPercent() throws IOException{
+    byte[] start = "aa".getBytes();
+    byte[] end = "zz".getBytes();
+    CloudBigtableScanConfiguration config = createDefaultConfig().withKeys(start, end).build();
+    CloudBigtableIO.Reader<Result> underTest = initializeReader(config);
+    ByteKeyRangeTracker tracker =
+        ByteKeyRangeTracker.of(ByteKeyRange.of(ByteKey.copyFrom(start), ByteKey.copyFrom(end)));
+
+    testTrackerAtKey(underTest, tracker, "bb", 1);
+    testTrackerAtKey(underTest, tracker, "qq", 2);
+
+    double splitAtFraction =
+        (1 - tracker.getFractionConsumed()) * .5 + tracker.getFractionConsumed();
+    ByteKey newSplitEnd = config.toByteKeyRange().interpolateKey(splitAtFraction);
+
+    underTest.splitAtFraction(splitAtFraction);
+    tracker.trySplitAtPosition(newSplitEnd);
+
+    Assert.assertEquals(tracker.getFractionConsumed(), underTest.getFractionConsumed(), 0.0001d);
+  }
+
+  private void testTrackerAtKey(CloudBigtableIO.Reader<Result> underTest, ByteKeyRangeTracker tracker,
+      final String key, final int count) throws IOException {
+    setRowKey(key);
+    tracker.tryReturnRecordAt(false, ByteKey.copyFrom(key.getBytes()));
+    Assert.assertTrue(underTest.start());
+    Assert.assertEquals(count, underTest.getRowsReadCount());
+    Assert.assertEquals(tracker.getFractionConsumed(),
+      underTest.getFractionConsumed().doubleValue(), .001d);
+  }
```

Comment:
the golden standard for correctness and thread-safety uses the `splitAtFractionExhaustive` test. It spins up multiple threads and tries to read all records and split in parallel to expose threading bugs.

https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/test/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIOTest.java#L295